objects/src/mock/chain.rs

```diff
@@ -55,3 +55,247 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+        },
+        Account, AccountError, AccountId, AccountType, Asset, AssetError, BlockHeader, ChainMmr,
+        Felt, FungibleAsset, NonFungibleAsset, StorageItem, Vec,
+    };
+    use crypto::{
+        hash::rpo::RpoDigest as Digest,
+        merkle::{DefaultMerkleStore, TieredSmt},
+    };
+    use miden_core::FieldElement;
+    use rand::{Rng, SeedableRng};
+
+    /// Initial timestamp value
+    const TIMESTAMP_START: Felt = Felt::new(1693348223);
+    /// Timestep of timestamp on each new block
+    const TIMESTAMP_STEP: Felt = Felt::new(10);
+
+    #[derive(Default, Debug, Clone)]
+    pub struct Objects<R: Rng> {
+        accounts: Vec<Account>,
+        fungible_faucets: Vec<(AccountId, FungibleAssetBuilder)>,
+        nonfungible_faucets: Vec<(AccountId, NonFungibleAssetBuilder<R>)>,
+    }
+
+    impl<R: Rng> Objects<R> {
+        pub fn new() -> Self {
+            Self {
+                accounts: vec![],
+                fungible_faucets: vec![],
+                nonfungible_faucets: vec![],
+            }
+        }
+
+        /// Extend this instance with the data of `other`.
+        pub fn extend(&mut self, other: &mut Objects<R>) {
+            self.accounts.extend(other.accounts.drain(..));
+            self.fungible_faucets.extend(other.fungible_faucets.drain(..));
+            self.nonfungible_faucets.extend(other.nonfungible_faucets.drain(..));
+        }
+    }
+
+    /// Structure chain data, used to build necessary openings and to construct [BlockHeader]s.
+    #[derive(Debug, Clone)]
+    #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+    pub struct MockChain<R: Rng + SeedableRng> {
+        /// An append-only structure used to represent the history of blocks produced for this chain.
+        chain: ChainMmr,
+
+        /// Storage for the historical state of the chain.
+        ///
+        /// This store contains historical data for accounts, nullifiers, notes, and batches.
+        data: DefaultMerkleStore,
+
+        /// History of produced blocks.
+        blocks: Vec<BlockHeader>,
+
+        /// Tree containing the latest `Nullifier`'s tree.
+        nullifier_db: TieredSmt,
+
+        // RNG used to seed builders.
+        //
+        // This is used to seed the [AccountBuilder] and the [NonFungibleAssetBuilder].
+        rng: R,
+
+        // Builder for new [AccountId]s of faucets.
+        account_id_builder: AccountIdBuilder<R>,
+
+        // Objects that have been created and commited to a block.
+        //
+        // These can be used to perform additional operations on a block.
+        #[cfg_attr(feature = "serde", serde(skip))]
+        objects: Objects<R>,
+
+        // Objects that have been created and are waiting for a block.
+        //
+        // These objects will become available once the block is sealed.
+        #[cfg_attr(feature = "serde", serde(skip))]
+        pending_objects: Objects<R>,
```

Comment:
skipping these because of the comment  https://github.com/0xPolygonMiden/miden-base/pull/211#discussion_r1308445952

> Having PendingBlock be a part of this struct feels a bit weird to me. When this struct is instantiated from a file, I'm not sure we need to have a "pending block". So, I wonder if this part can be refactored so that we do block construction externally.